apps/dash-medical-provider-charges/app.py

```diff
@@ -0,0 +1,736 @@
+                            values=["All"],
+                        ),
+                    ),
+                    html.Div(
+                        id="region-select-dropdown-outer",
+                        children=dcc.Dropdown(
+                            id="region-select",
+                            options=[{"label": i, "value": i} for i in region_list],
+                            value=region_list[:4],
+                            multi=True,
+                            searchable=True,
+                        ),
+                    ),
+                ],
+            ),
+        ],
+    )
+
+
+def generate_geo_map(geo_data, selected_metric, region_select, procedure_select):
+    filtered_data = geo_data[
+        geo_data["Hospital Referral Region (HRR) Description"].isin(region_select)
+    ]
+
+    colors = ["#21c7ef", "#76f2ff", "#ff6969", "#ff1717"]
+
+    hospitals = []
+
+    lat = filtered_data["lat"].tolist()
+    lon = filtered_data["lon"].tolist()
+    average_covered_charges_mean = filtered_data[selected_metric]["mean"].tolist()
+    regions = filtered_data["Hospital Referral Region (HRR) Description"].tolist()
+    provider_name = filtered_data["Provider Name"].tolist()
+
+    # Cost metric mapping from aggregated data
+
+    cost_metric_data = {}
+    cost_metric_data["min"] = filtered_data[selected_metric]["mean"].min()
+    cost_metric_data["max"] = filtered_data[selected_metric]["mean"].max()
+    cost_metric_data["mid"] = (cost_metric_data["min"] + cost_metric_data["max"]) / 2
+    cost_metric_data["low_mid"] = (
+        cost_metric_data["min"] + cost_metric_data["mid"]
+    ) / 2
+    cost_metric_data["high_mid"] = (
+        cost_metric_data["mid"] + cost_metric_data["max"]
+    ) / 2
+
+    for i in range(len(lat)):
+        val = average_covered_charges_mean[i]
+        region = regions[i]
+        provider = provider_name[i]
+
+        if val <= cost_metric_data["low_mid"]:
+            color = colors[0]
+        elif cost_metric_data["low_mid"] < val <= cost_metric_data["mid"]:
+            color = colors[1]
+        elif cost_metric_data["mid"] < val <= cost_metric_data["high_mid"]:
+            color = colors[2]
+        else:
+            color = colors[3]
+
+        selected_index = []
+        if provider in procedure_select["hospital"]:
+            selected_index = [0]
+
+        hospital = go.Scattermapbox(
+            lat=[lat[i]],
+            lon=[lon[i]],
+            mode="markers",
+            marker=dict(
+                color=color,
+                showscale=True,
+                colorscale=[
+                    [0, "#21c7ef"],
+                    [0.33, "#76f2ff"],
+                    [0.66, "#ff6969"],
+                    [1, "#ff1717"],
+                ],
+                cmin=cost_metric_data["min"],
+                cmax=cost_metric_data["max"],
+                size=10
+                * (1 + (val + cost_metric_data["min"]) / cost_metric_data["mid"]),
```

Comment:
Can you explain why you used this formula?

Reply:
A size calibration for individual scatters. should look into https://plot.ly/python/bubble-charts/#scaling-the-size-of-bubble-charts for a more standard workaround.

Reply:
the original tableau displayed size of scatter on map in relationship to its cost (not sure whether diameter or area). This formula here plots all scatters in current trace with a normalized diameter.  
In this case setting a sizeref like https://plot.ly/python/reference/#scatter-marker-sizeref may not be necessary since I just need a representation of relative value difference between all displayed marker.